probe-rs/src/architecture/arm/sequences/stm32.rs

```diff
@@ -0,0 +1,106 @@
+//! Sequences for STM32 devices
+
+use std::sync::Arc;
+
+use super::ArmDebugSequence;
+use crate::{
+    architecture::arm::{
+        ap::MemoryAp, ApAddress,
+        DpAddress,
+    },
+    Memory,
+};
+
+use anyhow::anyhow;
+
+/// Marker struct indicating initialization sequencing for STM32H7 family parts.
+pub struct Stm32h7 {}
+
+impl Stm32h7 {
+    /// Create the sequencer for the H7 family of parts.
+    pub fn create() -> Arc<Self> {
+        Arc::new(Self {})
+    }
+
+    /// Enable all debug components on the chip.
+    pub fn enable_debug_components(&self, memory: &mut Memory<'_>) -> Result<(), crate::Error> {
+        log::info!("Enabling STM32H7 debug components");
+
+        let mut control = dbgmcu::Control::read(memory)?;
+
+        // Enable the debug clock doamins for D1 and D3. This ensures we can access CoreSight
+        // components in these power domains.
+        control.enable_d1_clock(true);
+        control.enable_d3_clock(true);
+
+        // The TRACECK also has to be enabled to communicate with the TPIU.
+        control.enable_traceck(true);
+
+        // Enable debug connection in all power modes.
+        control.enable_standby_debug(true);
+        control.enable_sleep_debug(true);
+        control.enable_stop_debug(true);
+
+        control.write(memory)?;
+
+        Ok(())
+    }
+}
+
+mod dbgmcu {
+    use crate::Memory;
+    use bitfield::bitfield;
+
+    /// The base address of the DBGMCU component
+    const DBGMCU: u64 = 0xE00E_1000;
+
+    bitfield! {
+        pub struct Control(u32);
+        impl Debug;
+
+        pub u8, dbgsleep_d1, enable_sleep_debug: 0;
+        pub u8, dbgstop_d1, enable_stop_debug: 1;
+        pub u8, dbgstby_d1, enable_standby_debug: 2;
+
+        pub u8, d3dbgcken, enable_d3_clock: 22;
+        pub u8, d1dbgcken, enable_d1_clock: 21;
+        pub u8, traceclken, enable_traceck: 20;
+    }
+
+    impl Control {
+        /// The offset of the Control register in the DBGMCU block.
+        const ADDRESS: u64 = 0x04;
+
+        pub fn read(memory: &mut Memory<'_>) -> Result<Self, crate::Error> {
+            let contents = memory.read_word_32(DBGMCU + Self::ADDRESS)?;
+            Ok(Self(contents))
+        }
+
+        pub fn write(&mut self, memory: &mut Memory<'_>) -> Result<(), crate::Error> {
+            memory.write_word_32(DBGMCU + Self::ADDRESS, self.0)
+        }
+    }
+}
+
+impl ArmDebugSequence for Stm32h7 {
+    fn debug_port_start(
```

Comment:
@Yatekii I'm struggling where this setup sequence should actually be executed.

As implemented, `debug_port_start` is never actually called when executing using an ST-link probe, so this code never runs.

However, this code cannot run on any future sequences (e.g. `debug_device_unlock`) because that sequence uses a `Memory` interface to the core, where the DBGMCU component isn't accessible.

One approach that _does_ work with the current design (but is sub optimal) is accessing the DBGMCU through the processor core AP0. The only change needed for this is to use a different address. However, that feels wrong since you're using the system bus instead of the dedicated debug bus.

Reply:
Perhaps `debug_device_unlock`'s signature should be modified to accept the raw probe instead of the `Memory` interface? That would allow us to pick the correct memory interface to use.

Reply:
Yeah, I did some rubber-duck debugging with myself and indeed changed the API of sequencing. My only concern currently is that this leaves all the SWV trace configuration enabled when the probe is disconnected. I've observed issues re-opening the probe on future connections in this configuration, and it's not clear to me why (I specifically see USB timeouts).

We may need to implement some kind of clean-up sequence as well to de-configure tracing.

Reply:
@ryan-summers When you say "probe is disconnected", I assume you mean as part of a graceful ending of the `probe_rs::Session`. If that is the case, can you use the [Session::drop](https://github.com/probe-rs/probe-rs/blob/0e7e242742757f5499f83156fc4e81a40bc5049c/probe-rs/src/session.rs#L500-L512) to disable SWV trace config?

Reply:
I resolved this by adding the `debug_core_stop()` sequence and modified the `Session` drop implementation to disable tracing on the target.

For the H7, this also handles deconfiguring the DBGMCU on shutdown to preserve device configuration.